remote-api/src/main/java/org/alfresco/rest/api/impl/rules/ActionParameterConverter.java

```diff
@@ -0,0 +1,132 @@
+import org.json.JSONArray;
+import org.json.JSONException;
+
+@Experimental
+public class ActionParameterConverter
+{
+    private final DictionaryService dictionaryService;
+    private final ActionService actionService;
+    private final NamespaceService namespaceService;
+
+    public ActionParameterConverter(DictionaryService dictionaryService, ActionService actionService,
+                                    NamespaceService namespaceService)
+    {
+        this.dictionaryService = dictionaryService;
+        this.actionService = actionService;
+        this.namespaceService = namespaceService;
+    }
+
+    Map<String, Serializable> getConvertedParams(Map<String, Serializable> params, String name) {
+        final Map<String, Serializable> parameters = new HashMap<>(params.size());
+        final ParameterizedItemDefinition definition = actionService.getActionDefinition(name);
+        if (definition == null)
+        {
+            throw new NotFoundException(NotFoundException.DEFAULT_MESSAGE_ID, new String[]{name});
+        }
+
+        for (Map.Entry<String, Serializable> param : params.entrySet())
+        {
+            final ParameterDefinition paramDef = definition.getParameterDefintion(param.getKey());
+            if (paramDef == null && !definition.getAdhocPropertiesAllowed())
+            {
+                throw new InvalidArgumentException(InvalidArgumentException.DEFAULT_MESSAGE_ID, new String[]{param.getKey(), name});
+            }
+            if (paramDef != null)
+            {
+                final QName typeQName = paramDef.getType();
+                parameters.put(param.getKey(), convertValue(typeQName, param.getValue()));
+            } else
+            {
+                parameters.put(param.getKey(), param.getValue().toString());
+            }
+        }
+        return parameters;
+    }
+
+    private Serializable convertValue(QName typeQName, Object propertyValue) throws JSONException
+    {
+        Serializable value;
+
+        final DataTypeDefinition typeDef = dictionaryService.getDataType(typeQName);
+        if (typeDef == null)
+        {
+            throw new NotFoundException(NotFoundException.DEFAULT_MESSAGE_ID, new String[]{typeQName.toPrefixString()});
+        }
+
+        if (propertyValue instanceof JSONArray)
+        {
+            final String javaClassName = typeDef.getJavaClassName();
+            try
+            {
+                Class.forName(javaClassName);
+            } catch (ClassNotFoundException e)
+            {
+                throw new DictionaryException("Java class " + javaClassName + " of property type " + typeDef.getName() + " is invalid", e);
+            }
+
+            final int length = ((JSONArray) propertyValue).length();
+            final List<Serializable> list = new ArrayList<>(length);
+            for (int i = 0; i < length; i++)
+            {
+                list.add(convertValue(typeQName, ((JSONArray) propertyValue).get(i)));
```

Comment:
JSONArrays are a pain to iterate through.

Reply:
Yes, they are, but at least for now I couldn't find a faster/simpler way to make all types of action mappings sustaining backwards compatibility.